module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,49 @@
+#include "compression.hpp"
+
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+CompressedPGMBitMap compressGrayscale(const PGMBitMap& img) {
+    CompressedPGMBitMap bitMap;
+    bitMap.reserve(height * width);
+
+    for (const auto& row : img) {
+        WidthType consecutiveCount = 0;
+
+        for (uint8_t i = 0; i < width; i++) {
+            consecutiveCount++;
+            if (i == width -1 or row[i+1] != row[i]){
+                bitMap.emplace_back(std::make_pair(row[i], consecutiveCount));
+                consecutiveCount = 0;
+            }
+        }
+    }
+    bitMap.shrink_to_fit();
+    return bitMap;
+}
+
+PGMBitMap decompressGrayscale(const CompressedPGMBitMap& compressedImg) {
+    PGMBitMap decompressed;
+
+    auto currentEmptySlot = decompressed[0].begin();
+
+    for (const auto& pair : compressedImg) {
+        auto placeToFillTo = std::next(currentEmptySlot, pair.second);
+        std::fill(currentEmptySlot, placeToFillTo, pair.first);
+        currentEmptySlot = placeToFillTo;
+    }
+
+    return decompressed;
+}
+
+void printMap(const PGMBitMap& img) {
+    std::stringstream ss;
+    for (const auto& row : img) {
+        for (const Pixel pixel : row) {
+            ss << std::setw(4) << std::setfill(' ') << std::to_string(pixel);
+        }
+        ss << '\n';
+    }
+    std::cout << ss.str();
+}
```

Comment:
That's not what we meant by printing image. We can barely see the ninja. 1 character should have width of 1. Just print it as an ASCII, not the number.

![Zrzut ekranu 2020-06-19 o 09 04 46](https://user-images.githubusercontent.com/4595135/85106131-
f8656380-b20b-11ea-9a22-61c951916c15.png)

Reply:
@ziobron I assumed that printMap is just utility function for us to test the code, and would not be checked. For me it was clearer to print the numbers, but ok, i changed it

Reply:
Ok, I thought that you implemented it to get the achievement "Artysta ASCII" :)